components/renku_data_services/connected_services/db.py

```diff
@@ -213,33 +225,36 @@ async def authorize_callback(
             if connection is None:
                 raise errors.Unauthorized(message="You do not have the required permissions for this operation.")
 
-            if next_url:
-                query = urlencode([("next_url", next_url)])
-                callback_url = f"{callback_url}?{query}"
-
             client = connection.client
             adapter = get_provider_adapter(client)
             client_secret = (
                 decrypt_string(self.encryption_key, client.created_by_id, client.client_secret)
                 if client.client_secret
                 else None
             )
+            extra_data = connection.extra_data or dict()
+            code_verifier = extra_data.get("code_verifier")
+            code_challenge_method = "S256" if code_verifier else None
             async with self.async_oauth2_client_class(
                 client_id=client.client_id,
                 client_secret=client_secret,
                 scope=client.scope,
                 redirect_uri=callback_url,
+                code_challenge_method=code_challenge_method,
                 state=connection.state,
             ) as oauth2_client:
-                token = await oauth2_client.fetch_token(adapter.token_endpoint_url, authorization_response=raw_url)
+                token = await oauth2_client.fetch_token(
+                    adapter.token_endpoint_url, authorization_response=raw_url, code_verifier=code_verifier
+                )
 
                 logger.info(f"Token for client {client.id} has keys: {", ".join(token.keys())}")
 
                 connection.token = self._encrypt_token_set(token=token, user_id=connection.user_id)
                 connection.state = None
                 connection.status = schemas.ConnectionStatus.connected
 
-                return token
+                extra_data = connection.extra_data or dict()
```

Comment:
I think you should remove the authentication-request specific data from the database here. I.e. the next_url and the code_verifier. What happens if the current request did not set a value for `next_url`? You will end up reusing the `next_url` value from the previous authentication request, right?